### PR TITLE
buildRustPackage: factor out setting up cargoBuildHook, add maturinBuildHook

### PR DESCRIPTION
**Pull request by danieldk**
###### Motivation for this change

Convert the build phase of `buildRustPackage` into the `cargoBuildHook` hook. Also add a `maturinBuildHook` and convert the derivations that use `maturin` to `buildPythonPackage` + `cargoSetupHook` + `maturinBuildHook`.

Additional comments:

- `buildRustPackage` has an API change: remove the `target` argument of `buildRustPackage`, the target should always be in sync with the C/C++ compiler that is used.    
- Gathering of binaries has moved from `buildPhase` to `installPhase`, this simplifies the hook and orders this functionality logically with the installation logic.

Todo:

- [x] Add documentation (will keep this a draft request until this is done).
- [ ] Read the changes once more to see if I missed something.

Related issues:

- #83614
- #112438

CC @FRidh @Mic92 @adisbladis @DavHau

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/wasmer/default.nix [hunk 1]
```diff
@@ -1,63 +1,32 @@
...
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    inherit src;
+    name = "${pname}-${version}";
+    hash = "sha256-txOOia1C4W+nsXuXp4EytEn82CFfSmiOYwRLC4WPImc=";
   };
 
-in
-buildPythonPackage rec {
-  inherit pname version;
+  nativeBuildInputs = with rustPlatform; [ cargoSetupHook maturinBuildHook ];
```
Comment (FRidh) [inline pkgs/development/python-modules/wasmer/default.nix:26]: `format = "pyproject";` is missing

**Review by FRidh — COMMENTED** [review]
Nice work!

File: pkgs/development/python-modules/wasmer/default.nix (hunk 1 above)
Comment (danieldk) [reply pkgs/development/python-modules/wasmer/default.nix:28]: Thanks! The other packages are in pyproject format as well, so I updated them as well.

**Comment by danieldk** [comment]
Added documentation now as well and lifted the Draft PR toggle.

**Review by jonringer — APPROVED** [review]
diff LGTM

**Review by 0x4A6F — APPROVED** [review]

**Comment by danieldk** [comment]
@mic92 is this ok for you to merge?

**Comment by Mic92** [comment]
Result of `nixpkgs-review pr 112804` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>python3Packages.johnnycanencrypt</li>
    <li>python3Packages.retworkx</li>
    <li>python3Packages.wasmer</li>
  </ul>
</details>

**Comment by danieldk** [comment]
I just noticed that I made an error -- finding libraries and executables must be done before `checkPhase`. But I'll do a PR for `cargoInstallHook` tomorrow, which also fixes this issue.

**Comment by collares** [comment]
Is there a chance that this broke gnome-tour on master? When built via meson/ninja, it compiles resources (using `gnome.compile_resources`) before starting rust compilation, which then includes the generated file into a rust source using a macro (`include_bytes!`). On master, `cargoBuildHook` runs and builds the project itself before the resources are generated, leading to a compilation failure.

Useful links:
* https://nix-cache.s3.amazonaws.com/log/swga9wvr79ki8mmwwyk4qxzra0isb1ka-gnome-tour-3.38.0.drv
* https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gnome-tour/default.nix#L51

cc @worldofpeace

**Comment by vcunat** [comment]
In the meantime I really bisected the regression to the first commit of this PR (without understanding it).  A catch is that you need a later change
```diff
--- a/pkgs/development/libraries/webkitgtk/default.nix
+++ b/pkgs/development/libraries/webkitgtk/default.nix
@@ -155,6 +155,9 @@ stdenv.mkDerivation rec {
     "-DPORT=GTK"
     "-DUSE_LIBHYPHEN=OFF"
     "-DUSE_WPE_RENDERER=OFF"
+    # ensure backward compatibility with the latest version of icu:
+    # http://linuxfromscratch.org/blfs/view/svn/x/webkitgtk.html
+    "-DCMAKE_CXX_FLAGS=-DU_DEFINE_FALSE_AND_TRUE=1"
   ] ++ optionals stdenv.isDarwin [
     "-DENABLE_GRAPHICS_CONTEXT_3D=OFF"
     "-DENABLE_GTKDOC=OFF"
```
to fix the build of the webkit dependency (and some of these take a long time to build, too).

**Comment by collares** [comment]
Thanks for verifying! The above patch already landed in master I believe (d2aeeb0).

**Comment by vcunat** [comment]
Yes, it landed at once with the breakage during the last `staging-next` merge.  And now `gnome-tour` is fixed as well 🎉

**Pull request merged**